crates/wasm-shrink/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+    /// and input Wasm file.
+    pub fn new(predicate: impl Into<PathBuf>, input_wasm: impl Into<PathBuf>) -> WasmShrink {
+        WasmShrink {
+            output: None,
+            attempts: 1_000,
+            seed: 42,
+            log_level: LogLevel::Info,
+            predicate: predicate.into(),
+            input: input_wasm.into(),
+        }
+    }
+
+    /// The output file path to write the shrunken Wasm file to.
+    ///
+    /// By default, a file path based on the input will be generated.
+    pub fn output(mut self, output: impl Into<PathBuf>) -> WasmShrink {
+        self.output = Some(output.into());
+        self
+    }
+
+    /// The number of shrink attempts to try before considering a Wasm module as
+    /// small as it will ever get.
+    pub fn attempts(mut self, attempts: u32) -> WasmShrink {
+        self.attempts = attempts;
+        self
+    }
+
+    /// Run this configured Wasm shrinking task.
+    ///
+    /// Returns information and metrics about the shrink task, such as the
+    /// shrunken Wasm's output file path, the size of the input Wasm, the size
+    /// of the output Wasm, etc.
+    pub fn run(self) -> Result<ShrinkInfo> {
+        // Prerequisites for the predicate.
+        anyhow::ensure!(
+            self.predicate.is_file(),
+            "The predicate script '{}' does not exist.",
+            self.predicate.display()
+        );
+        anyhow::ensure!(
+            self.predicate.is_executable(),
+            "The predicate script '{}' is not executable.",
+            self.predicate.display()
+        );
+
+        // Prerequisites for the input Wasm.
+        anyhow::ensure!(
+            self.input.is_file(),
+            "The input Wasm file '{}' does not exist.",
+            self.input.display()
+        );
+        anyhow::ensure!(
+            self.is_valid_wasm(&self.input)?,
+            "The input Wasm file '{}' is not valid Wasm.",
+            self.input.display(),
+        );
+
+        let input_size = self.input.metadata()?.len();
+
+        let output = self
+            .output
+            .clone()
+            .unwrap_or_else(|| self.input.with_extension("shrunken.wasm"));
+        log::info!("Writing shrunken Wasm file to: {}", output.display());
+
+        // First double check that the input Wasm file passes the predicate.
+        //
+        // It is fairly common to accidentally write a predicate script that
+        // doesn't consider the input Wasm interesting. Better to surface this
+        // user error as quick as possible than to make them wait until we've
+        // exhausted all the ways we could shrink it further.
+        anyhow::ensure!(
+            self.test_file(&self.input)?,
+            "The predicate does not consider the input Wasm file interesting."
+        );
+
+        // The smallest Wasm that passes the predicate.
+        let mut best = vec![];
+        read_file(&self.input, &mut best)?;
```

Comment:
This is the only usage of `read_file` so having a special function that takes a destination-as-argument may not be too important? Could this use `std::fs::read`?